odxtools/datarecord.py

```diff
@@ -0,0 +1,73 @@
+# SPDX-License-Identifier: MIT
+from dataclasses import dataclass, field
+from typing import Any, cast
+from xml.etree import ElementTree
+
+from .datafile import Datafile
+from .dataformatselection import DataformatSelection
+from .element import NamedElement
+from .exceptions import odxraise, odxrequire
+from .identvalue import IdentValue
+from .odxdoccontext import OdxDocContext
+from .odxlink import OdxLinkDatabase, OdxLinkId
+from .snrefcontext import SnRefContext
+from .specialdatagroup import SpecialDataGroup
+from .utils import dataclass_fields_asdict
+
+
+@dataclass(kw_only=True)
+class DataRecord(NamedElement):
+    rule: str | None = None
+    key: str | None = None
+    data_id: IdentValue | None = None
+    sdgs: list[SpecialDataGroup] = field(default_factory=list)
+
+    # at most one of the following two attributes is not None
+    datafile: Datafile | None = None
+    data: str | None = None
```

Comment:
Maybe add helper method that returns actual data from either `data` or `datafile` ?

Reply:
I'm not so sure if having a non-trivial additional dependency is a good idea (this can quickly degenerate into dependency hell like most NPM and Java projects). also, `hexrec` seems to have problems with one of the datasets which have been used internally at our company for ages (it raises a `ValueError('non-contiguous data within range')` exception). `bincopy` can handle the dataset, but if fills these gaps in the record with `0xff` bytes instead of `0x00`. (That said, I'm not sure how "gaps" are supposed to be filled or if it even matters. I guess that using zeros might be advantageous, because the memory controller of the ECU's microchip might compress all-zero pages by pointing it to the same memory location...)